#### by Bernardo Vimpi
April 2020

### PROJECT DESCRIPTION AND SCOPE

In this project I incorporate Machine Learning technique using Sklearn along with Multiple Linear Regression to predict the GPA score of a given student based features: SAT Score and Rand, 1,2,3.

###### IMPORT THE NECESSARY LIBRARIES

In [49]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
from sklearn.linear_model import LinearRegression

####  DATA LOAD

In [50]:
data=pd.read_csv('original.csv')

In [51]:
data.head()

SAT  Rand 1,2,3   GPA
0  1714           1  2.40
1  1664           3  2.52
2  1760           3  2.54
3  1685           3  2.74
4  1693           2  2.83

#### DEFINE THE FEATURES AND VARIABLE

There are two imput features: SAT score and Rand 1,2,3. WHile the Dependednt Variable is GPA

#### DESCRIPTIVE STATISTICS OF THE DATA

In [52]:
data.describe()

SAT  Rand 1,2,3        GPA
count    84.000000   84.000000  84.000000
mean   1845.273810    2.059524   3.330238
std     104.530661    0.855192   0.271617
min    1634.000000    1.000000   2.400000
25%    1772.000000    1.000000   3.190000
50%    1846.000000    2.000000   3.380000
75%    1934.000000    3.000000   3.502500
max    2050.000000    3.000000   3.810000

#### CREATE THE MULTIPLE LINEAR REGRESSION

##### DEPENDENT AND INDEPENDENT VARIABLES

In [53]:
x= data[['SAT', 'Rand 1,2,3']]
y=data['GPA']

### REGRESSION 

In [54]:
reg =LinearRegression()
reg.fit(x,y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [55]:
reg.coef_

array([ 0.00165354, -0.00826982])

### Results and Interpretation
It displays the coefiecient of the SAT(0.0017)score and the coeficient of the Rand 1,2,3 variable (-0.00827)

###### INTERCEPT

In [56]:
reg.intercept_

0.29603261264909486

##### Results:
It shows us that the intecept of this model is 0.296

###### CALCULATING THE ADJUSTED-R SQUARED

The R-squared is a universal measure used to evaluate how well linear regressions fare and compare. 

In [57]:
reg.score(x,y)

0.4066811952814285

#### Results and Further Analysis:
The result above is the R-Squared. We need to calculate the adjusted-R-Squared. Since SKlean doesn't have a formula for it, we can use some mathematical computation to calculate the adjusted R-Squared

#### ADJUSTED R-Squared



$R^2_{adj.} = 1 - (1-R^2)*\frac{n-1}{n-p-1}$

In [58]:
x.shape

(84, 2)

##### Results
in the formula, n= number of obeservations or 84 while p is the number of predictors 2

In [59]:
r2= reg.score(x,y)
n=x.shape[0] # n=84
p= x.shape[1] # p=2
adjusted_r2 = 1 -(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.39203134825134023

#### Conclusion 1:

The adjusted R-Squared is considerably lower than the R-Squared. Therefore, one or more of the predictors have little or no explanatory power. If we can identify and remove such variable, we can remove it to improve our model. However, how do we go about detecting which variable has little to no predictive power. We use the process called "Feature Selection"

##### FEATURE SELECTION

Feature selection is a very important feature in machine learning. It aims to simplify models making them much easier to be interpreted by Data Scientists. Throug the process of feature selection we get improved speed and also prevent some unwanted issues arising from having too many features. In the Statsmodels, were also given the p-values of the features which are used to determine if the independent variables are relevant for the model. To expand on that, if a variable has a p-value > 0.05 we can disregard it. The next question is: How do we find the p-value in SKlearn? Klearn is a machine learning package and not a statistical one. So, there are no builtin methods we can call. That said, we have to use a workaraound. We can use the F-Regression which creates simple linear regressions of each feature and the dependent variable. In our case of GPA and SAT, it will translate into two regressions: One will predict GPA with SAT and the other will predict GPA with Rand 1,2,3. Thereafter, the method will calculate the F-Statistics for each of those regressions and output their respective P-values. If there were 100 features, 100 simple regressions will be created. It is important also to note that for a simple linear regression, the p-value of F-Statistics coinsides with the p-value of the only independent variable. Therefore, this is precisely the method we need here

In [60]:
#import the F_regression method
from sklearn.feature_selection import f_regression 

In [61]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

#### Results:
We do have two arrays. The first one contains the F-Statistics for each of the regressions while the second array contains the p-values. We are more intrested in the p-values. We'll work with that next.

In [62]:
# from the f_regressions, let's extract only the p-values and work with them
p_values = f_regression(x,y)[1]
p_values

array([7.19951844e-11, 6.76291372e-01])

#### LETS ROUND THE RESULTS OF THE P-VALUES

In [63]:
p_values.round(3)

array([0.   , 0.676])

The P-Value of SAT is 0.000 while the P-Value of Rand, 1,2,3 is 0.676. Based on this results from P-Values, we can conclude that SAT is a much useful variable while Rand 1,2,3 has no much power. However, it is worth to mention that these are the univariate p-values reached from simple linear models. They do not reflect the interconnection of the features in our multiple linear regression. Therefore, F_Regression should be used with a lot of caution as we move from a simplistic to a more complicated problem

#### SUMMARY TABLE OF THE DATA

In [64]:
#lets create a data frame
reg_summary= pd.DataFrame(data=['SAT', 'Rand,1,2,3'], columns=['Features'])
reg_summary

Features
0         SAT
1  Rand,1,2,3

In [65]:
reg_summary ['Coefficients']= reg.coef_
reg_summary ['p_values'] = p_values.round(3)
reg_summary

Features  Coefficients  p_values
0         SAT      0.001654     0.000
1  Rand,1,2,3     -0.008270     0.676

#### Conclusion 2:
Once again, we can conclude that Rand 1,2,3 does not contribute to our model and should be removed. Last but not least, P-Values are one of the best ways to determine if a variable is redundant or not. However, P-Values don't provide information on the usefulness of a variable. For example, two different variables may have the same P-Values of 0.000 making them, therefore, relevant. But this doesn't necessarely mean that both variables are equally important.

Thank you!

##### Bernardo Vimpi
April 2020